### DISCUSSION TITLE
caching asp.net

### DISCUSSION THREAD

**Discussion by habbassi869** [dotnet/aspnetcore#49772] [General]
hello 
i need help please  
    in this line  " if (!_memoryCache.TryGetValue(cacheKey, out ObjWrapper<ProfessionalHealthViewModel> cachedData))" when i test my fonction it shows me System.NullReferenceException : 'La référence d'objet n'est pas définie à une instance d'un objet 
        memoryCache is null

this is my fonction : 
 public class HealthCareProfessionalController : BaseController
    {
        private readonly IMemoryCache _memoryCache;

        public HealthCareProfessionalController()
        {
        }

        public HealthCareProfessionalController(AppUserManager userManager, IMemoryCache memoryCache)
        {
            _userManager = userManager;
            _memoryCache = memoryCache;
        }

        [Authorize(Roles = "PS_Update,HealthCareProfessionalBtn_Create_PS,Menu_SearchHealthCareProfessional")]
        public async Task<ActionResult> PrincipalHealthCareProfessionalForm(int? Id, bool isref = false)
        {
            string cacheKey = Id.HasValue ? "PSData_" + Id.ToString() : "PSData_Null";

            try
            {
                if (!_memoryCache.TryGetValue(cacheKey, out ObjWrapper<ProfessionalHealthViewModel> cachedData))
                {
                    try
                    {
                        var sw = Stopwatch.StartNew();
                        var vm = new ObjWrapper<ProfessionalHealthViewModel>();
                        var listTasks = new List<Task>();
                        Task<ListWrapper<SelectListItem>> task_DocumentsAgrement = null;
                        Task<ListWrapper<AlerteViewModel>> task_ListAlerte = null;
                        Task<ListWrapper<SelectListItem>> task_DefaultPays = null;

                        using (var client = GetClient())
                        {
                            ViewBag.TypeSaisie = GlobalVariables.TypeDeSaisie;
                            ViewBag.DefaultRegion = ReferencesService.Get(client, "DefaultRegion");

                            var task_TypePS = Task.Run(() => ReferentielsService.GetTypePS(client));
                            listTasks.Add(task_TypePS);

                            var task_TypecategoriesPS = Task.Run(() => ReferentielsService.GetTypeCategoriesPS(client));
                            listTasks.Add(task_TypecategoriesPS);

                            var task_countrySource = Task.Run(() => ReferencesService.Get(client, "Countries"));
                            listTasks.Add(task_countrySource);

                            var task_LanguesCommSource = Task.Run(() => TiersService.GetLanguesComm(client));
                            listTasks.Add(task_LanguesCommSource);

                            var task_PrestationsIntermediaires = Task.Run(() => ReferentielsService.GetPrestationsIntermediaires(client));
                            listTasks.Add(task_PrestationsIntermediaires);

                            if (Id.HasValue)
                            {
                                vm = ReferentielsService.GetPSById(client, Id);
                                if (vm != null && vm.Item != null)
                                {
                                    task_DocumentsAgrement = Task.Run(() => ReferentielsService.GetDocumentsAgrementPS(client, Id));
                                    listTasks.Add(task_DocumentsAgrement);

                                    task_ListAlerte = Task.Run(() => ReferentielsService.GetListAlertePs(client, (int)Id));
                                    listTasks.Add(task_ListAlerte);

                                    if (vm.Item.Address != null && vm.Item.Address.IdPays != null && vm.Item.Address.IdPays.HasValue)
                                    {
                                        // Faites quelque chose avec vm.Item.Address.IdPays
                                    }

                                    if (ViewBag.TypeSaisie == "domestique")
                                    {
                                        task_DefaultPays = Task.Run(() => ReferencesService.Get(client, "DefaultCountry"));
                                        listTasks.Add(task_DefaultPays);

                                        var task_LanguesCommSourceDefault = Task.Run(() => ReferencesService.Get(client, "DefaultLanguage"));
                                        listTasks.Add(task_LanguesCommSourceDefault);
                                    }
                                }
                                else
                                {
                                    return RedirectToAction("Error500", "Error", new { msg = "PS n'existe pas !!!" });
                                }
                            }
                            else
                            {
                                if (ViewBag.TypeSaisie == "domestique")
                                {
                                    task_DefaultPays = Task.Run(() => ReferencesService.Get(client, "DefaultCountry"));
                                    listTasks.Add(task_DefaultPays);
                                }
                            }

                            // Exécuter les tâches de manière asynchrone
                            await Task.WhenAll(listTasks.ToArray());

                            // Mettre les données en cache
                            _memoryCache.Set(cacheKey, vm, new MemoryCacheEntryOptions
                            {
                                AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(30)
                            });

                            // Vérifier si les tâches ont retourné des listes null et les remplacer par des listes vides
                            ViewBag.DocumentsAgrement = task_DocumentsAgrement?.Result ?? new ListWrapper<SelectListItem>();
                            ViewBag.TypePS = task_TypePS.Result;
                            ViewBag.TypecategoriesPS = task_TypecategoriesPS.Result;
                            ViewBag.countrySource = task_countrySource.Result;
                            ViewBag.LanguesCommSource = task_LanguesCommSource.Result;
                            ViewBag.PrestationsIntermediaires = task_PrestationsIntermediaires.Result;

                            if (ViewBag.TypeSaisie == "domestique")
                            {
                                ViewBag.DefaultPays = task_DefaultPays?.Result ?? new ListWrapper<SelectListItem>();
                            }

                            if (Id.HasValue)
                            {
                                ViewBag.IsEditPS = true;
                                var infos_administratives_PS = Task.Run(() => ReferentielsService.GetInformationsAdministrativesByPS(client, (int)Id));
                                var PS_prestation_intermediaire = Task.Run(() => ReferentielsService.GetPrestationsIntermediairesByPS(client, (int)Id));

                                await Task.WhenAll(infos_administratives_PS, PS_prestation_intermediaire);

                                sw.Stop();

                                vm.Item.infos_administratives_PSViewModel = infos_administratives_PS.Result.Items;
                                vm.Item.PS_prestation_intermediaireViewModel = PS_prestation_intermediaire.Result.Items;

                                ViewBag.time = "( Time taken : " + sw.Elapsed.ToString(@"mm\:ss") + " )";
                                var filename = "PrincipalHealthCareProfessional_TimeTaken_MAI";
                                if (ViewBag.TypeSaisie == "domestique")
                                {
                                    filename = "PrincipalHealthCareProfessional_TimeTaken_IGA";
                                }

                                ViewBag.host = CustomViewService.hostname_sagelia;
                                var message = DateTime.Now + ";" + sw.Elapsed.ToString(@"mm\:ss") + ";" + vm.Item.Id + ";" + ViewBag.host;
                                WriteToLogData(message, filename);
                                return View("~/Views/Referentiels/PS/PsForm.cshtml", vm.Item);
                            }
                            else
                            {
                                sw.Stop();
                                ViewBag.time = "( Time taken : " + sw.Elapsed.ToString(@"mm\:ss") + " )";
                                var filename1 = "PrincipalHealthCareProfessional_TimeTaken_MAI";
                                if (ViewBag.TypeSaisie == "domestique")
                                {
                                    filename1 = "PrincipalHealthCareProfessional_TimeTaken_IGA";
                                }
                                var vm1 = new ProfessionalHealthViewModel();
                                ViewBag.host = CustomViewService.hostname_sagelia;
                                var message1 = DateTime.Now + ";" + sw.Elapsed.ToString(@"mm\:ss") + ";" + vm1.Id + ";" + ViewBag.host;
                                WriteToLogData(message1, filename1);
                                return View("~/Views/Referentiels/PS/PsForm.cshtml", vm1);
                            }
                        }
                    }
                    catch (Exception ex)
                    {
                        // Gérer les exceptions potentielles ici
                        string errorMessage = $"Une exception s'est produite : {ex.Message}\nStackTrace: {ex.StackTrace}";
                        return RedirectToAction("Error500", "Error", new { msg = errorMessage });

                    }
                }
                else
                {
                    // Si les données sont déjà dans le cache, utilisez-les directement
                    return View("~/Views/Referentiels/PS/PsForm.cshtml", cachedData.Item);
                }
            }
            catch (Exception ex)
            {
                // Gérer les exceptions potentielles ici
                string errorMessage = $"Une exception s'est produite : {ex.Message}\nStackTrace: {ex.StackTrace}";
                return RedirectToAction("Error500", "Error", new { msg = errorMessage });

            }
        }

**Comment by LeaFrock** [comment]
> memoryCache is null

Have you ever injected `IMemoryCache`? It has nothing to do with the codes you pasted. 
You should check if a code line like `builder.Services.AddMemoryCache();` exists.
